server/odc-service/src/main/java/com/oceanbase/odc/service/iam/UserService.java

```diff
@@ -192,44 +190,55 @@ public class UserService {
                     .build(key -> new FailedLoginAttemptLimiter(FAILED_LOGIN_ATTEMPT_TIMES,
                             FAILED_LOGIN_ATTEMPT_LOCK_TIMEOUT));
 
+    /**
+     * Create user if not exists, or update user if extraPropertiesJson is different
+     * 
+     * @param userEntity user entity, must not be null
+     * @param roleNames role names, only for create new user
+     * @return created or updated user
+     */
     @SkipAuthorize("odc internal usage")
     @Transactional(rollbackFor = Exception.class)
-    public User createUserIfNotExists(Long organizationId, String accountName, String name, List<String> roleNames) {
-        Optional<UserEntity> userEntityOptional = userRepository.findByAccountName(accountName);
-        if (userEntityOptional.isPresent()) {
-            log.info("User already exists, accountName={}", accountName);
-            return new User(userEntityOptional.get());
-        }
-        UserEntity userEntity = initUser(organizationId, accountName, name);
-        userRepository.saveAndFlush(userEntity);
-        UserEntity userEntityCreated = userRepository.findByAccountName(userEntity.getAccountName()).get();
-        Long userId = userEntityCreated.getId();
-        User user = new User(userEntityCreated);
-        if (Objects.nonNull(roleNames)) {
-            List<Long> roleIds = new ArrayList<>();
-            List<Role> roles = new ArrayList<>();
-            for (String roleName : roleNames) {
-                Optional<RoleEntity> roleEntity = roleRepository.findByNameAndOrganizationId(roleName, organizationId);
-                PreConditions.validExists(ResourceType.ODC_ROLE, "roleName", roleName, roleEntity::isPresent);
-                PreConditions.validArgumentState(roleEntity.get().getType() != RoleType.INTERNAL,
-                        ErrorCodes.BadArgument,
-                        new Object[] {"Internal role does not allow operation"},
-                        "Internal role does not allow operation");
-                Long roleId = roleEntity.get().getId();
-                UserRoleEntity userRoleEntity = new UserRoleEntity();
-                userRoleEntity.setUserId(userId);
-                userRoleEntity.setRoleId(roleId);
-                userRoleEntity.setOrganizationId(organizationId);
-                userRoleEntity.setCreatorId(OdcConstants.DEFAULT_ADMIN_USER_ID);
-                userRoleRepository.saveAndFlush(userRoleEntity);
-
-                roleIds.add(roleId);
-                roles.add(new Role(roleEntity.get()));
+    public User createIfNotExistsOrUpdate(UserEntity userEntity, List<String> roleNames) {
```

Comment:
method name suggestion: createIfNotExistsOrUpdate --> upsert

Reply:
Good suggestion, I will modify it in the future.

Reply:
Since you renamed the method name in this PR, so why not just rename it to a more understandable name directly in this PR?

Reply:
Done.